lib/src/source_visitor.dart

```diff
@@ -886,17 +886,35 @@ class SourceVisitor extends ThrowingAstVisitor {
 
     // If the body is a binary operator expression, then we want to force the
     // split at `=>` if the operators split. See visitBinaryExpression().
-    if (node.expression is! BinaryExpression) builder.endRule();
+    var expression = node.expression;
+    if (expression is! BinaryExpression) builder.endRule();
 
     if (_isInLambda(node)) builder.endSpan();
 
-    builder.startBlockArgumentNesting();
+    // If the body is a block-formatted function invocation, don't add extra
+    // nesting to preserve normal indentation for decorators that modify the
+    // functionality of a method body:
+    //
+    //    void foo() => decorate(() {
+    //      "the body";
+    //    });
+    var isDecorator = expression is InvocationExpression &&
+        expression.argumentList.arguments.any((argument) {
```

Comment:
This logic is similar to, but not the same as the logic for handling indentation of function literals in other argument lists. In particular, I don't think you'd want to apply the special decorator logic if there are non-function arguments before and after the function in the function list, or between functions.

The logic for determining when functions can be un-indented in an argument list is pretty sensitive, so it would be good to reuse that code if at all possible.